### PR TITLE
feat: add release automation workflows

### LINKED ISSUE #343: feat: add release automation workflows for vote issue and draft release creation
## What would you like to be added?

Add two GitHub Actions workflows to automate the manual steps in the release process:

1. **`release-vote.yml`** (`workflow_dispatch`): Automates creation of the release vote issue. When triggered with a tag name (and optional commit SHA), it:
   - Dynamically parses `MAINTAINERS.md` to build a checkbox list of approvers
   - Extracts PR numbers from `git log` between the previous tag and the target commit
   - Creates a formatted vote issue with maintainer list, PR changelog, and full changelog link
   - Handles first-release edge case (no previous tag)

2. **`release-github.yml`** (tag push trigger): Automates creation of the GitHub draft release. When a `v*` tag is pushed, it:
   - Looks up PR metadata via `gh pr view` API with a fallback to commit message parsing when API lookup fails (e.g. on forks)
   - Categorizes PRs by conventional commit prefix into Breaking Changes, New Features, Bug Fixes, and Other Changes sections
   - Creates a **draft** release (never auto-publishes) to require manual review before publishing
   - Automatically detects pre-release versions (alpha, beta, rc, preview)
   - Includes NuGet package link and full changelog comparison URL

3. **`RELEASE_CHECKLIST.md`** updated to document the new automated flow.

### Design Decisions

- **Release steps don't change**: The overall release process remains the same — vote, tag, release, publish. These workflows replace the manual issue creation and release note authoring with automation.
- **Tag push remains manual**: Creating and pushing the tag is still a deliberate manual action by the release manager/maintainers. This preserves the human gate between a successful vote and the actual release.
- **Draft releases only**: `release-github.yml` always creates draft releases, never auto-publishing. This ensures maintainers can review and edit release notes before making them public.
- **Fallback mechanism for PR metadata**: When `gh pr view` fails (e.g. API rate limits, fork testing), the workflow falls back to parsing commit messages for PR title and author. Fallback authors use plain text (no `@` prefix) to avoid tagging the wrong person.
- **Conventional commit categorization**: PRs are categorized using title prefixes (`feat:`, `fix:`, `feat!:`, etc.) matching the project's existing commit convention, producing structured release notes consistent with the [v0.4.0 release](https://github.com/oras-project/oras-dotnet/releases/tag/v0.4.0) format.

## Why is this needed for the ORAS .NET library?

The current release process requires manually creating vote issues and drafting release notes, which is error-prone and time-consuming. The [existing RELEASE_CHECKLIST.md](https://github.com/oras-project/oras-dotnet/blob/main/RELEASE_CHECKLIST.md) describes several manual steps that can be automated while preserving the same deliberate release flow. This is consistent with automation in sibling ORAS projects (e.g. [oras-go](https://github.com/oras-project/oras-go)).

### PR DESCRIPTION
**Pull request by akashsinghal**
### What this PR does / why we need it

Adds two GitHub Actions workflows to automate manual steps in the release process, replacing error-prone manual issue creation and release note drafting with automation. **The release steps themselves don't change** — this PR automates the execution of existing steps while preserving the same deliberate flow.

Additionally, `release-nuget.yml` is updated to trigger on `release: published` instead of `push: tags: v*`, ensuring NuGet packages are only published after a maintainer explicitly publishes the draft release. The workflow also uploads the `.nupkg` and a SHA256 checksum as release assets for transparency and offline verification.

### Release process with this PR

1. **Determine version** — pick a SemVer2 tag (e.g., `v1.0.0-rc.1`)
2. **Run `release-vote` workflow** — from the Actions tab, provide the tag name and optional commit SHA. A vote issue is created with maintainer checklist and PR changelog.
3. **Wait for vote to pass** — maintainers approve via LGTM on the issue.
4. **Push the tag** — `git tag <tag_name> <commit_sha> && git push origin <tag_name>`. This creates a **draft** GitHub Release with auto-generated structured notes.
5. **Review the draft** — edit release notes on the Releases page if needed.
6. **Publish the release** — this triggers NuGet package build, publish to nuget.org, and uploads the `.nupkg` + SHA256 checksum as release assets.
7. **Wait for NuGet validation** and announce.

### Validation

All three workflows were end-to-end tested on a [personal fork](https://github.com/akashsinghal/oras-dotnet). The full flow was validated multiple times across iterations:

1. **`release-vote`** — Triggered via `workflow_dispatch` with `v0.7.0-test`. Vote issue created with correct maintainer count (4), checkbox list parsed from `MAINTAINERS.md`, PR changelog populated, tag name validation passed. First-release edge case (no previous tags) handled correctly.
2. **`release-github`** — Tag push created a draft release with categorized notes (Breaking Changes, New Features, Bug Fixes, Other Changes, Detailed Commits). Pre-release detection, tag validation, and `git tag --merged` ancestry all verified.
3. **`release-nuget`** — Triggered only on release publish (not on tag push). Build succeeded, `.nupkg` and `.sha256` checksum uploaded as release assets **before** the NuGet push step. NuGet push failed as expected (no API key configured on fork), confirming that assets are attached even if NuGet push encounters a transient failure. `--skip-duplicate` ensures re-runs are safe.

### What changed

**`release-vote.yml`** (`workflow_dispatch`):
- Triggered manually from the Actions tab with a tag name and optional commit SHA
- Validates tag name against SemVer2 format and commit SHA with `git rev-parse --verify`
- Dynamically parses `MAINTAINERS.md` to build a checkbox list of approvers
- Counts maintainers via `grep -c` on checkbox lines for accuracy
- Uses `git tag --merged` to find previous tag reachable from the target commit
- Extracts PR numbers from `git log` between the previous tag and the target commit
- Creates a formatted vote issue with maintainer list, PR changelog, and full changelog link
- Handles first-release edge case when no previous tag exists

**`release-github.yml`** (tag push trigger):
- Triggered automatically when a `v*` tag is pushed — tag push remains a deliberate manual action by the release manager so maintainers control when a release happens
- Validates tag name against SemVer2 format
- Uses `git tag --merged` to find previous tag reachable from the current tag (correct for release branches)
- Looks up PR metadata via `gh pr view` API with a fallback to commit message parsing when API lookup fails
- Falls back to raw commit listing when no PR numbers are found (direct commits)
- Categorizes PRs by conventional commit prefix into Breaking Changes, New Features, Bug Fixes, and Other Changes — consistent with the [v0.4.0 release notes](https://github.com/oras-project/oras-dotnet/releases/tag/v0.4.0) format
- **Always creates draft releases** (never auto-publishes) to require manual review before making release notes public
- Automatically detects pre-release versions (alpha, beta, rc, preview) and marks them accordingly
- Includes NuGet package link and full changelog comparison URL
- Permissions: `contents: write` for release creation, `pull-requests: read` for PR metadata lookup

**`release-nuget.yml`** (updated trigger + release assets):
- Changed from `push: tags: v*` to `release: types: [published]`
- Validates tag name against SemVer2 format before version extraction
- NuGet package is now only published when a maintainer explicitly publishes the draft release
- Prevents accidental package publication before release notes are reviewed
- **Uploads `.nupkg` and SHA256 checksum as release assets before NuGet push** — assets survive transient push failures
- Uses `--skip-duplicate` on NuGet push for safe re-runs
- Permissions: `contents: write` for release asset uploads

**`RELEASE_CHECKLIST.md`** updated to document the new automated flow including release asset uploads.

### Key design decisions

- **Draft-only releases**: Ensures maintainers review notes before publishing. No accidental public releases.
- **NuGet gated on publish**: NuGet package is only published after the draft release is explicitly published, not on tag push. This allows retracting a draft without any package being published.
- **Release asset uploads**: The `.nupkg` and its SHA256 checksum are attached to the GitHub release for archival, transparency, and offline verification. Assets are uploaded before NuGet push so they persist even if the push fails. NuGet.org remains the primary distribution channel.
- **Manual tag push**: Preserves the human gate between a successful vote and the actual release.
- **Tag ancestry via `git tag --merged`**: Previous tag is resolved relative to the target commit, ensuring correct behavior for both linear releases and release branches/backports.
- **SemVer2 tag validation**: All three workflows validate the tag name format, guarding against injection and malformed input.
- **PR metadata fallback**: When `gh pr view` fails, falls back to parsing commit messages. Fallback authors use plain text (no `@`) to avoid tagging the wrong person. When no PRs are found at all, falls back to raw commit listing.
- **`grep -xFv` for tag filtering**: Uses whole-line fixed-string matching to avoid regex metacharacter issues with version dots.
- **First-release handling**: Both workflows guard against empty previous tag, using full `git log` and omitting changelog comparison URL when appropriate.
- **Minimal permissions**: Each workflow declares only the permissions it needs (`contents: read/write`, `issues: write`, `pull-requests: read`).

### Which issue(s) this PR resolves / fixes

Fixes #343

### Please check the following list
- [x] Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [x] Does this change require a documentation update?
- [ ] Does this introduce breaking changes that would require an announcement or bumping the major version?
- [x] Do all new files have an appropriate license header?

### CODE REVIEW

File: .github/workflows/release-github.yml [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            # First release: include all commits
+            PR_NUMBERS=$(git log --oneline "$TAG_NAME" \
+              | grep -oP '#\K\d+' | sort -n -u || true)
+          fi
+
+          # Build a map of PR numbers to commit message data
+          # for fallback when API lookup fails.
+          declare -A COMMIT_MAP
+          while IFS= read -r line; do
+            if [[ "$line" =~ \#([0-9]+)\) ]]; then
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-github.yml]: Good observation. The \#([0-9]+)\) pattern is intentionally strict — it matches the standard GitHub squash-merge format Title (#123) which is consistent across all commits in this repo (verified across the full commit history).

This regex is only used for building the COMMIT_MAP fallback (when gh pr view API fails). The primary PR number extraction on line 55 uses the looser #\K\d+ pattern which matches #123 regardless of surrounding characters. So even if a commit didn't have the closing paren, the PR would still be discovered — it just wouldn't have a fallback entry, which is a graceful degradation rather than a silent miss.

The strict pattern here avoids false positives in the fallback map (e.g., matching #123 in the middle of a sentence that isn't a PR reference).

File: .github/workflows/release-github.yml [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            if [ -n "$PR_DATA" ]; then
+              PRS="${PRS}${PR_DATA}"$'\n'
+            elif [ -n "${COMMIT_MAP[$num]:-}" ]; then
+              # Fallback: parse title and author from commit msg.
+              # Author is plain text (no @) to avoid wrong tags.
+              RAW="${COMMIT_MAP[$num]}"
+              TITLE=$(echo "$RAW" \
+                | sed -E "s/ *\(#${num}\)\t.*$//" )
+              AUTHOR=$(echo "$RAW" \
+                | sed -E 's/^.*\t//')
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-github.yml]: This sed pattern is in the fallback path, which only executes when the gh pr view API lookup fails. The input format is the GitHub squash-merge default Title (#123)\tAuthorName (tab-separated via git log --format="%s%x09%an"), which is consistent across all commits in this repo.

Regarding $num — it's always a numeric value extracted by grep -oP '#\K\d+' upstream, so there's no risk of special characters or injection.

That said, if the extraction fails to match, TITLE would retain the full string including the (#NNN) suffix, which would result in a slightly redundant but still correct release note entry (the PR number would appear in both the title and the in #NNN suffix). It degrades gracefully rather than producing incorrect output.

File: .github/workflows/release-vote.yml [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+          # Build changelog context
+          if [ -n "$PREV_TAG" ]; then
+            CHANGES_HEADER="The changes compared to \`${PREV_TAG}\` include:"
+            CHANGELOG_LINK="See [the full changelog](https://github.com/${REPO}/compare/${PREV_TAG}...${COMMIT_SHA}) for more details."
+          else
+            CHANGES_HEADER="The changes in this release include:"
+            CHANGELOG_LINK=""
+          fi
+
+          BODY="At least 2 approvals are needed from the ${COUNT} maintainers for tagging ${COMMIT_SHA} as \`${TAG_NAME}\`."
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-vote.yml]: will not split in this case since for scripts like this in a block, it will make it difficult to read

File: .github/workflows/release-vote.yml [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+          PREV_TAG="${{ steps.prev_tag.outputs.tag }}"
+          MAINTAINERS="${{ steps.maintainers.outputs.list }}"
+          PR_LIST="${{ steps.changelog.outputs.pr_list }}"
+          REPO="${{ github.repository }}"
+          COUNT=$(echo "$MAINTAINERS" | wc -l)
+
+          # Build changelog context
+          if [ -n "$PREV_TAG" ]; then
+            CHANGES_HEADER="The changes compared to \`${PREV_TAG}\` include:"
+            CHANGELOG_LINK="See [the full changelog](https://github.com/${REPO}/compare/${PREV_TAG}...${COMMIT_SHA}) for more details."
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-vote.yml]: will not fix. for bash scripts like this in a block of text in yaml, splitting across 2 lines will make readability worse

File: .github/workflows/release-github.yml [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+          fi
+
+          NOTES="${NOTES}"$'\n'"## Detailed Commits"$'\n\n'"${COMMITS}"
+          if [ -n "$PREV_TAG" ]; then
+            CHANGELOG="https://github.com/${REPO}/compare/${PREV_TAG}...${TAG_NAME}"
+            NOTES="${NOTES}"$'\n'"**Full Changelog**: [${PREV_TAG}...${TAG_NAME}](${CHANGELOG})"
+          fi
+
+          # Write notes to file for gh release
+          echo "$NOTES" > release-notes.md
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-github.yml]: fixed

File: .github/workflows/release-nuget.yml [hunk 6]
```diff
@@ -30,8 +29,25 @@ jobs:
...
         run: dotnet nuget push ./src/OrasProject.Oras/bin/Release/OrasProject.Oras.${{ steps.version.outputs.version }}.nupkg --source https://api.nuget.org/v3/index.json --api-key ${{ secrets.NUGET_API_KEY }}
+      - name: Upload nupkg and checksum to release
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          NUPKG: ./src/OrasProject.Oras/bin/Release/OrasProject.Oras.${{ steps.version.outputs.version }}.nupkg
+        run: |
+          sha256sum "$NUPKG" | awk '{print $1}' > "${NUPKG}.sha256"
+          gh release upload "${{ github.event.release.tag_name }}" \
+            "$NUPKG" \
+            "${NUPKG}.sha256"
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-nuget.yml]: fixed

File: .github/workflows/release-vote.yml [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        required: true
+        type: string
+      commit_sha:
+        description: >
+          The commit SHA to tag for this release.
+          Defaults to the latest commit on the default branch.
+        required: false
+        type: string
+
+permissions:
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-vote.yml]: fixed

File: .github/workflows/release-vote.yml [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+            fi
+            echo "sha=${SHA}" >> "$GITHUB_OUTPUT"
+          else
+            echo "sha=$(git rev-parse HEAD)" >> "$GITHUB_OUTPUT"
+          fi
+
+      - name: Get previous release tag
+        id: prev_tag
+        run: |
+          PREV_TAG=$(git tag --sort=-v:refname | head -n 1)
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-vote.yml]: fixed

File: .github/workflows/release-vote.yml [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+  contents: read
+  issues: write
+
+jobs:
+  create-vote-issue:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v6
+        with:
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-vote.yml]: fixed

File: .github/workflows/release-nuget.yml [hunk 10]
```diff
@@ -30,8 +32,25 @@ jobs:
...
         run: dotnet build ./src/OrasProject.Oras --configuration Release /p:PackageVersion=${{ steps.version.outputs.version }}
+      - name: Upload nupkg and checksum to release
+        env:
+          GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          NUPKG: ./src/OrasProject.Oras/bin/Release/OrasProject.Oras.${{ steps.version.outputs.version }}.nupkg
+        run: |
+          sha256sum "$NUPKG" | awk '{print $1}' > "${NUPKG}.sha256"
+          gh release upload "${{ github.event.release.tag_name }}" \
+            "$NUPKG" \
+            "${NUPKG}.sha256"
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-nuget.yml]: fixed

File: .github/workflows/release-github.yml [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v6
+        with:
+          fetch-depth: 0
+
+      - name: Validate tag name
+        run: |
+          if ! echo "${GITHUB_REF_NAME}" \
+            | grep -qP '^v\d+\.\d+\.\d+(-[a-zA-Z0-9.\-]+)?(\+[a-zA-Z0-9.\-]+)?$'; then
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-github.yml]: Fair point that the regex doesn't enforce every SemVer2 rule (leading zeros, empty identifiers). However, the
validation is intended as a guard against injection and obviously malformed input, not a full SemVer2 parser. Tags
are pushed by maintainers after a vote, so there's already a human review gate. A canonical SemVer2 regex would be significantly more complex and harder to maintain in shell. The current pattern strikes a practical balance — it
catches injection payloads and clearly wrong formats while remaining readable.

File: .github/workflows/release-vote.yml [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+          fetch-depth: 0
+
+      - name: Validate tag name
+        run: |
+          TAG_NAME="${{ inputs.tag_name }}"
```
Comment (shizhMSFT) [inline .github/workflows/release-vote.yml]: `${{ inputs.tag_name }}` is directly interpolated into the shell by GitHub's template engine _before_ the script runs. This is vulnerable to [script injection](https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions#understanding-the-risk-of-script-injections) — a crafted input can execute arbitrary commands.

We should pass all user-controlled inputs via `env:` mapping:
```yaml
      - name: Validate tag name
        env:
          TAG_NAME: ${{ inputs.tag_name }}
        run: |
          if ! echo "$TAG_NAME" ...
```
Same pattern applies to all `${{ inputs.* }}` and `${{ steps.*.outputs.* }}` expressions used directly in `run:` blocks across all three workflow files.

File: .github/workflows/release-github.yml [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+            NOTES="${NOTES}"$'\n'"## New Features"$'\n\n'"${FEATURES}"
+          fi
+          if [ -n "$FIXES" ]; then
+            NOTES="${NOTES}"$'\n'"## Bug Fixes"$'\n\n'"${FIXES}"
+          fi
+          if [ -n "$OTHER" ]; then
+            NOTES="${NOTES}"$'\n'"## Other Changes"$'\n\n'"${OTHER}"
+          fi
+
+          NOTES="${NOTES}"$'\n'"## Detailed Commits"$'\n\n'"${COMMITS}"
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-github.yml]: Renamed to All Changes to clarify its intentionally a flat list.

File: .github/workflows/release-github.yml [hunk 14]
```diff
@@ -0,0 +1,208 @@
...
+
+          # Make release creation idempotent: edit if release exists, otherwise create.
+          if gh release view "$TAG_NAME" \
+            --repo "$REPO" >/dev/null 2>&1; then
+            gh release edit "$TAG_NAME" \
+              --repo "$REPO" \
+              --title "$TAG_NAME" \
+              --notes-file release-notes.md \
+              --draft \
+              $PRERELEASE_FLAG
```
Comment (akashsinghal) [inline .github/workflows/release-github.yml]: nit: `\$PRERELEASE_FLAG` is intentionally unquoted to expand to nothing when empty, but ShellCheck (SC2086) flags this. Same on line 207.

File: RELEASE_CHECKLIST.md [hunk 15]
```diff
@@ -7,11 +7,16 @@ This document describes the checklist to publish a release through the [GitHub r
...
-4. Compose and revise the release note and optionally select `Set as a pre-release` depending on the version.
-5. Publish the release on GitHub.
-6. A [workflow](https://github.com/oras-project/oras-dotnet/actions/workflows/release-nuget.yml) will be triggered automatically by tag creation mentioned in the step 3 for publishing the release to NuGet.
+2. Run the [`release-vote`](.github/workflows/release-vote.yml) workflow from the Actions tab. Provide the tag name and optionally a commit SHA (defaults to latest on `main`). This creates a vote issue listing maintainers and the changelog since the last release (see [example](https://github.com/oras-project/oras-dotnet/issues/103)).
+3. After the vote passes, push the tag targeting the voted commit:
+   ```bash
+   git tag <tag_name> <commit_sha>
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply RELEASE_CHECKLIST.md]: Reworded to focus on backport/release-branch guidance.

File: .github/workflows/release-github.yml [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+              --notes-file release-notes.md \
+              --draft \
+              ${PRERELEASE_FLAG:+"$PRERELEASE_FLAG"}
+          else
+            gh release create "$TAG_NAME" \
+              --repo "$REPO" \
+              --title "$TAG_NAME" \
+              --notes-file release-notes.md \
+              --draft \
+              ${PRERELEASE_FLAG:+"$PRERELEASE_FLAG"}
```
[the remark this answers is not in the document]
Comment (akashsinghal) [reply .github/workflows/release-github.yml]: Fixed -- the edit path now queries the existing release state via gh release view --json isDraft and only passes --draft when the release is already a draft. New releases are still always created as drafts.

**Comment by akashsinghal** [comment]
@sajayantony this PR is ready to merge. I've validated release process once again after the rounds of automated PR review and updates

**Review by sajayantony — APPROVED** [review]
